base/scopedvalues.jl

```diff
@@ -0,0 +1,178 @@
+    implementation is available from the package ScopedValues.jl.
+"""
+mutable struct ScopedValue{T}
+    const initial_value::T
+end
+
+Base.eltype(::Type{ScopedValue{T}}) where {T} = T
+
+##
+# Notes on the implementation.
+# We want lookup to be unreasonably fast.
+# - IDDict/Dict are ~10ns
+# - ImmutableDict is faster up to about ~15 entries
+# - ScopedValue are meant to be constant, Immutabilty
+#   is thus a boon
+# - If we were to use IDDict/Dict we would need to split
+#   the cache portion and the value portion of the hash-table,
+#   the value portion is read-only/write-once, but the cache version
+#   would need a lock which makes ImmutableDict incredibly attractive.
+#   We could also use task-local-storage, but that added about 12ns.
+# - Values are GC'd when scopes become unreachable, one could use
+#   a WeakKeyDict to also ensure that values get GC'd when ScopedValues
+#   become unreachable.
+# - Scopes are an inline implementation of an ImmutableDict, if we wanted
+#   be really fancy we could use a CTrie or HAMT.
+
+mutable struct Scope
+    const parent::Union{Nothing, Scope}
+    const key::ScopedValue
+    const value::Any
+    Scope(parent, key::ScopedValue{T}, value::T) where T = new(parent, key, value)
+end
+Scope(parent, key::ScopedValue{T}, value) where T =
+    Scope(parent, key, convert(T, value))
+
+function Scope(scope, pairs::Pair{<:ScopedValue}...)
+    for pair in pairs
+        scope = Scope(scope, pair...)
+    end
+    return scope
+end
+
+"""
+    current_scope()::Union{Nothing, Scope}
+
+Return the current dynamic scope.
+"""
+current_scope() = current_task().scope::Union{Nothing, Scope}
+
+function Base.show(io::IO, scope::Scope)
+    print(io, Scope, "(")
+    seen = Set{ScopedValue}()
+    while scope !== nothing
+        if scope.key ∉ seen
+            if !isempty(seen)
+                print(io, ", ")
+            end
+            print(io, typeof(scope.key), "@")
+            show(io, Base.objectid(scope.key))
+            print(io, " => ")
+            show(IOContext(io, :typeinfo => eltype(scope.key)), scope.value)
+            push!(seen, scope.key)
+        end
+        scope = scope.parent
+    end
+    print(io, ")")
+end
+
+function Base.getindex(var::ScopedValue{T})::T where T
+    scope = current_scope()
+    while scope !== nothing
+        if scope.key === var
+            return scope.value::T
+        end
+        scope = scope.parent
+    end
+    return var.initial_value
+end
+
+function Base.show(io::IO, var::ScopedValue)
+    print(io, ScopedValue)
+    print(io, '{', eltype(var), '}')
+    print(io, '(')
+    show(IOContext(io, :typeinfo => eltype(var)), var[])
+    print(io, ')')
+end
+
+"""
+    with(f, (var::ScopedValue{T} => val::T)...)
+
+Execute `f` in a new scope with `var` set to `val`.
+"""
+function with(f, pair::Pair{<:ScopedValue}, rest::Pair{<:ScopedValue}...)
+    @nospecialize
+    ct = Base.current_task()
```

Comment:
`current_task()` is called repeatedly when entering a scope with multiple assignments. Does that have a substantive performance impact?

Reply:
Also the point of this method is to only cal it once? We don't call `with` recursively